libs/grok/src/main/java/org/elasticsearch/grok/PatternBank.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.grok;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class PatternBank {
+
+    public static PatternBank EMPTY = new PatternBank(Map.of());
+
+    private final Map<String, String> bank;
+
+    public PatternBank(Map<String, String> bank) {
+        Objects.requireNonNull(bank, "bank must not be null");
+        forbidCircularReferences(bank);
+
+        // the bank reference should be unmodifiable, based on a defensive copy of the passed-in bank, and
+        // maintain the iteration order of the passed-in bank (assuming there was a meaningful order)
+        this.bank = Collections.unmodifiableMap(new LinkedHashMap<>(bank));
+    }
+
+    public String get(String patternName) {
+        return bank.get(patternName);
+    }
+
+    public Map<String, String> bank() {
+        return bank;
+    }
+
+    /**
+     * Extends a pattern bank with extra patterns, returning a new pattern bank.
+     * <p>
+     * The returned bank will be the same reference as the original pattern bank if the extra patterns map is null or empty.
+     *
+     * @param extraPatterns the patterns to extend this bank with (may be empty or null)
+     * @return the extended pattern bank
+     */
+    public PatternBank extendWith(Map<String, String> extraPatterns) {
+        if (extraPatterns == null || extraPatterns.isEmpty()) {
+            return this;
+        }
+
+        var extendedBank = new LinkedHashMap<>(bank);
+        extendedBank.putAll(extraPatterns);
+        return new PatternBank(extendedBank);
+    }
+
+    /**
+     * Checks whether patterns reference each other in a circular manner and if so fail with an exception.
+     * <p>
+     * In a pattern, anything between <code>%{</code> and <code>}</code> or <code>:</code> is considered
+     * a reference to another named pattern. This method will navigate to all these named patterns and
+     * check for a circular reference.
+     */
+    static void forbidCircularReferences(Map<String, String> bank) {
+        // first ensure that the pattern bank contains no simple circular references (i.e., any pattern
+        // containing an immediate reference to itself) as those can cause the remainder of this algorithm
+        // to recurse infinitely
+        for (Map.Entry<String, String> entry : bank.entrySet()) {
+            if (patternReferencesItself(entry.getValue(), entry.getKey())) {
```

Comment:
Wait, never mind. This way looks better since it gives a simpler error message for self referencing patterns. Can we update the comment here to mention that? Looks like this was just a move of some previous code, but the old comment could use a quick update.